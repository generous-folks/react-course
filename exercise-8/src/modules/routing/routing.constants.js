export const ROUTES_PATHS_BY_NAMES = {
  home: '/',
  login: '/login',
  about: '/about',
  contact: '/contact',
  article: '/articles/:id',
  checkout: '/checkout',
};

export const PROTECTED_PATHS = [ROUTES_PATHS_BY_NAMES.login, ROUTES_PATHS_BY_NAMES.checkout];
