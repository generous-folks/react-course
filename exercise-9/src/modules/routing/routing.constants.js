export const ROUTES = {
  home: '/',
  login: '/login',
  about: '/about',
  contact: '/contact',
  article: '/articles/:id',
  checkout: '/checkout',
};

export const PROTECTED_ROUTES = [ROUTES.checkout];
