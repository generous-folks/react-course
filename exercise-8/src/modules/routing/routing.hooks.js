import { useEffect, useState, useMemo } from 'react';
import { useUserState } from '../user/user.context';

import { useHistory, useLocation } from 'react-router-dom';
import { PROTECTED_ROUTES, ROUTES } from './routing.constants';
import { isUserConnected } from '../user/user.selectors';

export const useLoginRedirect = () => {
  const state = useUserState();
  const isConnected = isUserConnected(state);
  const { pathname } = useLocation();
  const { push } = useHistory();

  const [initialRoute, setInitialRoute] = useState(
    pathname === ROUTES.login ? ROUTES.home : pathname,
  );

  const isProtectedRoute = PROTECTED_ROUTES.includes(pathname);

  const isLoginRoute = useMemo(() => pathname === ROUTES.login, [pathname]);
  const isSameLocation = useMemo(() => pathname === initialRoute, [pathname, initialRoute]);
  const redirectRouteDestination = useMemo(() => initialRoute, [initialRoute]);

  useEffect(() => {
    if (isConnected && isLoginRoute && !isSameLocation) {
      push(redirectRouteDestination);
    }
  }, [isConnected, push, isLoginRoute, isSameLocation, redirectRouteDestination]);

  useEffect(() => {
    if (!isConnected && pathname !== ROUTES.login && isProtectedRoute) {
      setInitialRoute(pathname);
      push(ROUTES.login);
    }
  }, [isConnected, push, pathname, isProtectedRoute]);
};
