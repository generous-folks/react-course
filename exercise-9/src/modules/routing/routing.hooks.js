import { useEffect, useState, useMemo } from 'react';
import { useHistory, useLocation } from 'react-router-dom';

import { useUserState } from '../user/user.context';
import { PROTECTED_ROUTES, ROUTES } from './routing.constants';
import { isUserConnected } from '../user/user.selectors';

export const useLoginRedirect = () => {
  const state = useUserState();
  const isConnected = isUserConnected(state);
  const { pathname } = useLocation();
  const { push } = useHistory();

  const [initialRoute, setInitialRoute] = useState(
    pathname === ROUTES.login ? ROUTES.home : pathname,
  );

  const isProtectedRoute = PROTECTED_ROUTES.includes(pathname);
  const isLoginRoute = useMemo(() => pathname === ROUTES.login, [pathname]);

  useEffect(() => {
    if (isConnected && isLoginRoute) {
      push(initialRoute);
    }
  }, [isConnected, push, isLoginRoute, initialRoute]);

  useEffect(() => {
    if (!isConnected && isProtectedRoute) {
      setInitialRoute(pathname);
      push(ROUTES.login);
    }
  }, [isConnected, push, pathname, isProtectedRoute]);
};
