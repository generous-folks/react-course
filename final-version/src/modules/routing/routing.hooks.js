import { useEffect, useState, useMemo } from 'react';
import { useHistory, useLocation } from 'react-router-dom';

import { useUserState } from '../user/user.context';
import { PROTECTED_PATHS, ROUTES_PATHS_BY_NAMES } from './routing.constants';
import { isUserConnected } from '../user/user.selectors';

const { login: loginPath, home: homePath } = ROUTES_PATHS_BY_NAMES;

export const useLoginRedirect = () => {
  const state = useUserState();
  const isConnected = isUserConnected(state);
  const { pathname } = useLocation();
  const { push } = useHistory();

  const [initialRoute, setInitialRoute] = useState(pathname === loginPath ? homePath : pathname);

  const isProtectedRoute = PROTECTED_PATHS.includes(pathname);
  const isLoginRoute = useMemo(() => pathname === loginPath, [pathname]);

  useEffect(() => {
    if (isConnected && isLoginRoute) {
      push(initialRoute);
    }
  }, [isConnected, push, isLoginRoute, initialRoute]);

  useEffect(() => {
    if (!isConnected && isProtectedRoute) {
      setInitialRoute(pathname);
      push(loginPath);
    }
  }, [isConnected, push, pathname, isProtectedRoute]);
};
