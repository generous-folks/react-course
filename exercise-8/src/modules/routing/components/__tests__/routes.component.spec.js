import React from 'react';

import { shallow } from 'enzyme';
import { AppRoutes } from '../routes.component';
import { ROUTES } from '../../routing.constants';
import { Route } from 'react-router-dom';

jest.mock('../../routing.hooks.js', () => ({ useLoginRedirect: jest.fn() }));

let wrapper;

describe('<AppRoutes />', () => {
  describe('Snapshot', () => {
    it('should render correctly', () => {
      expect(shallow(<AppRoutes />)).toMatchSnapshot();
    });
  });

  describe('Markup checks', () => {
    beforeEach(() => {
      wrapper = shallow(<AppRoutes />);
    });

    it('should have routes for all paths', () => {
      expect(
        wrapper
          .find(Route)
          .map(node => node.prop('path'))
          .sort(),
      ).toMatchObject(Object.values(ROUTES).sort());
    });
  });
});
