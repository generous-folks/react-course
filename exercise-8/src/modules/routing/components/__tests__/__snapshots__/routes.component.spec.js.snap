// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AppRoutes /> Snapshot should render correctly 1`] = `
<Fragment>
  <Switch>
    <Route
      path="/about"
    >
      <AboutPage />
    </Route>
    <Route
      path="/contact"
    >
      <ContactPage />
    </Route>
    <Route
      path="/login"
    >
      <LoginPage />
    </Route>
  </Switch>
  <ArticlesProvider>
    <CartProvider>
      <Switch>
        <Route
          exact={true}
          path="/checkout"
        >
          <CheckoutPage />
        </Route>
        <Route
          path="/articles/:id"
        >
          <ArticlePage />
        </Route>
        <Route
          exact={true}
          path="/"
        >
          <HomePage />
        </Route>
      </Switch>
    </CartProvider>
  </ArticlesProvider>
</Fragment>
`;
